lib/ecto/changeset.ex

```diff
@@ -2067,6 +2065,124 @@ defmodule Ecto.Changeset do
+      iex> changeset.valid?
+      false
+
+  """
+  @spec validate_any_required(t, list | atom, Keyword.t) :: t
+  def validate_any_required(%Changeset{} = changeset, fields, opts \\ []) when not is_nil(fields) do
+    %{validations: validations, errors: errors, types: types} = changeset
+    fields = List.wrap(fields)
+
+    valid_field =
+      Enum.reduce_while(fields, nil, fn field, acc ->
+        if not invalid_required_field!(changeset, types, field),
+          do: {:halt, field},
+          else: {:cont, acc}
+      end)
+
+    if valid_field do
+      new_validation = {valid_field, {:any_required, fields: fields}}
+      %{changeset | validations: [new_validation | validations]}
+    else
+      message = message(opts, "at least one field must be present")
+      new_error = {hd(fields), {message, [validation: :any_required, fields: fields]}}
+      %{changeset | errors: [new_error | errors], valid?: false}
+    end
+  end
+
+  @doc """
+  Validates exactly one of the fields is present in the changeset.
+
+  See `validate_required/3` for more details on what constitues a present field.
+  Note that unlike `validate_required/3`, this function will store its validations
+  under the `changeset.validations` key instead of the `changeset.required` key.
+  It will also add an error even if another one already exists on the field.
+
+  ## Options
+
+    * `:message` - the message on failure, defaults to "exactly one field must be present"
+
+  ## Examples
+
+      iex> changeset = cast(%Post{}, %{title: "Title"}, [:title])
+      iex> changeset = validate_one_of_required(changeset, [:title, :body])
+      iex> changeset.valid?
+      true
+
+      iex> changeset = cast(%Post{}, %{title: "Title", body: "Body"}, [:title, :body])
+      iex> changeset = validate_one_of_required(changeset, [:title, :body])
+      iex> changeset.valid?
+      false
+
+      iex> changeset = cast(%Post{}, %{color: "Red"}, [:color])
+      iex> changeset = validate_one_of_required(changeset, [:title, :body])
+      iex> changeset.valid?
+      false
+
+  """
+  @spec validate_one_of_required(t, list | atom, Keyword.t) :: t
+  def validate_one_of_required(%Changeset{} = changeset, fields, opts \\ []) when not is_nil(fields) do
+    %{validations: validations, errors: errors, types: types} = changeset
+    fields = List.wrap(fields)
+
+    valid_fields =
+      Enum.reduce_while(fields, [], fn field, acc ->
+        acc =
+          if not invalid_required_field!(changeset, types, field),
+            do: [field | acc],
+            else: acc
+
+        case acc do
+          [_, _] -> {:halt, acc}
+          _ -> {:cont, acc}
+        end
+      end)
+
+    case valid_fields do
+      [valid_field] ->
+        new_validation = {valid_field, {:one_of_required, fields: fields}}
+        %{changeset | validations: [new_validation | validations]}
+
+      _ ->
+        error_field = if valid_fields == [], do: hd(fields), else: hd(valid_fields)
```

Comment:
I think it would likely be easier to manage by callers if the field we add the error is always the same (i.e. the first field).